### ISSUE TITLE
Decode from indices directly

### ISSUE THREAD

**Issue by alexanderswerdlow** [zhaoyue-zephyrus/bsq-vit#4]
Hi,

Is there a code sample where you decode directly from indices (instead of just zq)? I tried using `get_codebook_entry`, but was receiving some very weird results.

Thanks!

**Comment by zhaoyue-zephyrus** [comment]
Hi @alexanderswerdlow ,

Could you provide more details on how are you using `get_codebook_entry`? The tokenizer decoder does expect zq as input to generate appropriate reconstruction result. The indices are only useful if you want to train a generation model on top of the tokens.

Best,
Yue

**Comment by alexanderswerdlow** [comment]
Hi @zhaoyue-zephyrus 

Among several other things, I was trying:
```
quant = vae.quantize.get_codebook_entry(indices)
image = vae.decode(quant)
```

I was going to try to train a model on top of them so I was seeing if this would be possible.

**Comment by zhaoyue-zephyrus** [comment]
This looks good to me.

Can you try the following:
```python
quant = vae.quantize.get_codebook_entry(indices)
image = vae.decode(quant)
image = vae.clamp(image)
```

The decoder does not guarantee the output is in proper range (even it's trained to do so). Clamping ensures the range is correct. (in our case it's -1, 1)

**Comment by alexanderswerdlow** [comment]
Yep! That's exactly what I was trying but I was getting some very weird results. Below is a zoomed in screenshot. Full encode/decode worked fine however.

<img width="979" alt="image" src="https://github.com/user-attachments/assets/af162c38-f4e3-4e6d-90f8-7c8f2809a77c">

![33](https://github.com/user-attachments/assets/7a53ed6c-d862-49a9-82d6-5d1f1220952a)

**Comment by VJJJJJJ1** [comment]
> This looks good to me.
> 
> Can you try the following:
> 
> quant = vae.quantize.get_codebook_entry(indices)
> image = vae.decode(quant)
> image = vae.clamp(image)
> The decoder does not guarantee the output is in proper range (even if it's trained to do so). Clamping ensures the range is correct. (in our case it's -1, 1)

but how to predict token using llm or small transformer when codebook size is large ? (eg. 2^18 or 2^20)

**Comment by VJJJJJJ1** [comment]
> This looks good to me.
> 
> Can you try the following:
> 
> quant = vae.quantize.get_codebook_entry(indices)
> image = vae.decode(quant)
> image = vae.clamp(image)
> The decoder does not guarantee the output is in proper range (even if it's trained to do so). Clamping ensures the range is correct. (in our case it's -1, 1)

Is there any sample about  how to predict token using llm or small transformer when codebook size is large ? (eg. 2^18 or 2^20)
